crates/integrations/datafusion/src/physical_plan/project.rs

```diff
@@ -0,0 +1,661 @@
+};
+use futures::StreamExt;
+use iceberg::spec::{PartitionSpec, Schema};
+
+use crate::to_datafusion_error;
+
+/// Prefix for partition column names to avoid collisions with regular columns
+const PARTITION_COLUMN_PREFIX: &str = "__partition_";
+
+/// An execution plan node that calculates partition values for Iceberg tables.
+///
+/// This execution plan takes input data from a child execution plan and adds partition columns
+/// based on the table's partition specification. The partition values are computed by applying
+/// the appropriate transforms to the source columns.
+///
+/// The output schema includes all original columns plus additional partition columns.
+#[derive(Debug, Clone)]
+pub(crate) struct IcebergProjectExec {
+    input: Arc<dyn ExecutionPlan>,
+    partition_spec: Arc<PartitionSpec>,
+    table_schema: Arc<Schema>,
+    output_schema: ArrowSchemaRef,
+    plan_properties: PlanProperties,
+}
+
+/// IcebergProjectExec is responsible for calculating partition values for Iceberg tables.
+/// It takes input data from a child execution plan and adds partition columns based on the table's
+/// partition specification. The partition values are computed by applying the appropriate transforms
+/// to the source columns. The output schema includes all original columns plus additional partition
+/// columns.
+impl IcebergProjectExec {
+    pub fn new(
+        input: Arc<dyn ExecutionPlan>,
+        partition_spec: Arc<PartitionSpec>,
+        table_schema: Arc<Schema>,
+    ) -> DFResult<Self> {
+        let output_schema =
+            Self::create_output_schema(&input.schema(), &partition_spec, &table_schema)?;
+        let plan_properties = Self::compute_properties(&input, output_schema.clone());
+
+        Ok(Self {
+            input,
+            partition_spec,
+            table_schema,
+            output_schema,
+            plan_properties,
+        })
+    }
+
+    /// Compute the plan properties for this execution plan
+    fn compute_properties(
+        input: &Arc<dyn ExecutionPlan>,
+        schema: ArrowSchemaRef,
+    ) -> PlanProperties {
+        PlanProperties::new(
+            EquivalenceProperties::new(schema),
+            input.output_partitioning().clone(),
+            EmissionType::Incremental,
+            Boundedness::Bounded,
+        )
+    }
+
+    /// Create the output schema by adding partition columns to the input schema
+    fn create_output_schema(
+        input_schema: &ArrowSchema,
+        partition_spec: &PartitionSpec,
+        table_schema: &Schema,
+    ) -> DFResult<ArrowSchemaRef> {
+        if partition_spec.is_unpartitioned() {
+            return Ok(Arc::new(input_schema.clone()));
+        }
+
+        let mut fields: Vec<Arc<Field>> = input_schema.fields().to_vec();
+
+        let partition_struct = partition_spec
+            .partition_type(table_schema)
+            .map_err(to_datafusion_error)?;
+
+        for (idx, pf) in partition_spec.fields().iter().enumerate() {
```

Comment:
Should we just have a new column with hardcoded name like `_iceberg_partition_values`? 

I think this way it can be easier accessed by the following operators like repartition node. Otherwise we will still need to combine the values together later to get a joined partition values to sort/repartition by